File: 21.c
#include <stdio.h>
#include <stdlib.h>

struct ListNode {
	int val;
	struct ListNode *next;
};
struct ListNode* mergeTwoLists(struct ListNode* l1, struct ListNode* l2) {
	int isToLow = 0;
	struct ListNode* result = NULL;
    if (l1 == NULL && l2 == NULL) {
    	return NULL;
    }

    if (l1 == NULL && l2 != NULL) {
    	return l2;
    }

    if (l1 != NULL && l2 == NULL) {
    	return l1;
    }

    if (l1->next != NULL) {
    	isToLow = l1->val > l1->next->val ? 1 : 0;
    } else if (l2->next != NULL) {
    	isToLow = l2->val > l2->next->val ? 1 : 0;
    }

    if (isToLow) {
    	while(l1 != NULL && l2 != NULL) {
    		if (result == NULL)  {
    			if (l1->val > l2->val) {
    				result = l1;
    				l1 = l1->next;
    			} else {
    				result = l2;
    				l2 = l2->next;
    			}
    			continue;
    		}

    		if (l1 == NULL) {
    			result->next = l2;
    			return result;
    		}

    		if (l2 == NULL) {
    			result->next = l1;
    			return result;
    		}

    		if (l1->val > l2->val) {
    			result->next = l1;
    			l1 = l1->next;
    			result = result->next;
    		}
    		if (l1->val < l2->val) {
    			result->next = l2;
    			l2 = l2->next;
    			result = result->next;
    		}
    	}
    }

    if (!isToLow) {
    	while(l1 != NULL && l2 != NULL) {
    		if (result == NULL)  {
    			if (l1->val < l2->val) {
    				result = l1;
    				l1 = l1->next;
    			} else {
    				result = l2;
    				l2 = l2->next;
    			}
    			continue;
    		}

    		if (l1 == NULL) {
    			result->next = l2;
    			return result;
    		}

    		if (l2 == NULL) {
    			result->next = l1;
    			return result;
    		}
    		if (l1->val < l2->val) {
    			result->next = l1;
    			l1 = l1->next;
    			result = result->next;
    		}
    		if (l1->val > l2->val) {
    			result->next = l2;
    			l2 = l2->next;
    			result = result->next;
    		}
    	}
    }
    return result;
}

int main(void)
{
	struct ListNode *l1 = (struct ListNode*) malloc(sizeof(struct ListNode));
	l1->val = -9;
	l1->next = (struct ListNode*) malloc(sizeof(struct ListNode));
	l1->next->val = 3;
	l1->next->next = NULL;
	struct ListNode *l2 = (struct ListNode*) malloc(sizeof(struct ListNode));
	l2->val = 5;
	l2->next = (struct ListNode*) malloc(sizeof(struct ListNode));
	l2->next->val = 7;
	l2->next->next = NULL;

	mergeTwoLists(l1, l2);
	return 0;
}

File: 21.java
class ListNode {
     int val;
     ListNode next;
     ListNode(int x) { val = x; }
}

class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode result = null;
        ListNode temp = null;
        if (l1 == null && l2 == null) {
            return null;
        }

        if (l1 == null) {
            return l2;
        }

        if (l2 == null) {
            return l1;
        }

        while (l1 != null && l2 != null) {
            if (l1.val < l2.val) {
                if (result == null) {
                    result = l1;
                    temp = result;
                } else {
                    temp.next = l1;
                    temp = temp.next;
                }
                l1 = l1.next;
            } else if (l2.val < l1.val) {
                if (result == null) {
                    result = l2;
                    temp = result;
                } else {
                    temp.next = l2;
                    temp = temp.next;
                }
                l2 = l2.next;
            } else if (l1.val == l2.val) {
                if (result == null) {
                    result = new ListNode(l1.val);
                    temp = result;
                    temp.next = new ListNode(l2.val);
                    temp = temp.next;
                } else {
                    temp.next = new ListNode(l1.val);
                    temp = temp.next;
                    temp.next = new ListNode(l2.val);
                    temp = temp.next;
                }
                l1 = l1.next;
                l2 = l2.next;
            }
        }

        if (l1 == null && l2 != null) {
            temp.next = l2;
        }

        if (l2 == null && l1 != null) {
            temp.next = l1;
        }
        return result;
    }
}